### PR TITLE
QuadTree implementation

### PR DESCRIPTION
**Pull request by akrammon**
QuadTree implementation and testing.

### CODE REVIEW

File: src/Core/Indexes/QuadTree.cs [hunk 1]
```diff
@@ -0,0 +1,400 @@
+﻿namespace AEGIS.Indexes
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    public class QuadTree : ISpatialIndex
```
Comment (robertogiachetta) [inline src/Core/Indexes/QuadTree.cs]: Please comment on types.

File: src/Core/Indexes/QuadTree.cs [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    {
+        protected class QuadTreeNode
+        {
+            /// <summary>
+            /// The region of the node.
+            /// </summary>
+            private Envelope bound;
```
Comment (robertogiachetta) [inline src/Core/Indexes/QuadTree.cs]: The "bound" name and the "region" description are misleading, suggest using "envelope" instead.

File: src/Core/Indexes/QuadTree.cs [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+
+            /// <summary>
+            /// The children nodes of this node. Each internal node must always have 4 children.
+            /// </summary>
+            private List<QuadTreeNode> children = new List<QuadTreeNode>(4);
```
Comment (robertogiachetta) [inline src/Core/Indexes/QuadTree.cs]: Field initialization should be performed in a constructor. Also, if the field is not replaced after the constructor, using the `readonly` modifier makes the code more efficient.

File: src/Core/Indexes/QuadTree.cs [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+                get
+                {
+                    return this.contents.Count == 0 && this.IsLeaf;
+                }
+            }
+
+            /// <summary>
+            /// Gets the contents of this node.
+            /// </summary>
+            /// <value> The geometries stored in this node.</value>
```
Comment (robertogiachetta) [inline src/Core/Indexes/QuadTree.cs]: Please check for spaces before and after comment, as they are taken up by VS. I.e. it should be "<value>The geometries ".

File: src/Core/Indexes/QuadTree.cs [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        /// <summary>
+        /// The root of the tree.
+        /// </summary>
+        private QuadTreeNode root;
+
+        /// <summary>
+        /// Gets a value indicating whether the index is read-only.
+        /// </summary>
+        /// <value><c>true</c> if the index is read-only; otherwise, <c>false</c>.</value>
+        public Boolean IsReadOnly => false;
```
Comment (robertogiachetta) [inline src/Core/Indexes/QuadTree.cs]: I suggest to be consistent on expression-bodied members, so either use them with all properties, or use the regular property format. (I myself prefer the regular fashion, but I'm probably just getting old.)

File: src/Core/Indexes/QuadTree.cs [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        /// <param name="geometries"> The geometries to add to the tree. </param>
+        public QuadTree(IEnumerable<IBasicGeometry> geometries)
+        {
+            List<Envelope> envelopes = new List<Envelope>();
+            foreach (IBasicGeometry geometry in geometries)
+            {
+                envelopes.Add(geometry.Envelope);
+            }
+
+            Envelope bound = Envelope.FromEnvelopes(envelopes);
```
Comment (robertogiachetta) [inline src/Core/Indexes/QuadTree.cs]: You can use the `Select` method instead of creating a list of envelopes.

**Comment by akrammon** [comment]
I don't know what the issue is/was with the csproj file, I couldn't get git to notice any differences. It's a mystery to me. Does it work now?

**Comment by robertogiachetta** [comment]
My mistake on the csproj file, the csproj format does not include source files anymore, because they are automatically built as long as they are in the project directory.

**Pull request merged**